# EDA of clean, wide-span data

## Imports

In [11]:
import pandas as pd
import re
from collections.abc import Iterable
pd.set_option("display.max_columns", None)

## Constants

In [12]:
MODULE_1 = [
    "GH 1",
    "GH 2",
    "GH 3",
    "GH 4",
    "GH 5",
    "GH 6",
    "GH 7",
    "GH 8",
    "GH 9",
    "GH 10",
    "GH 11",
    "GH 12",
    "GH 13",
    "GH 14",
    "GH 15",
    "GH 17",
    "GH 19",
    "GH 21",
    "GH 22",
]

MODULE_2 = [
    "GH 24",
    "GH 25",
    "GH 27",
    "GH 28",
    "GH 29",
    "GH 30",
    "GH 31",
    "GH 32",
    "GH 33",
    "GH 34",
    "GHT",
]

MODULE_3 = ["IF 1", "IF 2", "IF 3", "IF 4", "IF 5", "IF 6", "IFT"]

MODULE_6 = ["AN 1", "AN 2", "AN 3", "AN 4", "AN 5", "AN 6", "AN 7"]

LESSONS_TO_KEEP = MODULE_1 + MODULE_2 + MODULE_3

In [13]:
FEATURE_SUFFIXES = [
    "score",
    "adjscore",
    "failed",
    "missing",
]

## Helper functions

In [14]:
def summarise_col(df, col):
    counts = df[col].value_counts(dropna=False)
    pct = df[col].value_counts(dropna=False, normalize=True).mul(100)
    summary = pd.DataFrame({
        "count": counts,
        "pct": pct.round(2)
    })
    print(summary)

In [15]:
def natural_sort_key(s):
    parts = re.split(r"(\d+)", str(s))
    return [int(p) if p.isdigit() else p for p in parts]

## Load data

In [16]:
df = pd.read_parquet("../data/03_clean/bwc_160_207.parquet")

In [17]:
df

,STUDENT_ID,GH 1_score,GH 1_adjscore,GH 1_failed,GH 1_missing,GH 2_score,GH 2_adjscore,GH 2_failed,GH 2_missing,GH 3_score,GH 3_adjscore,GH 3_failed,GH 3_missing,GH 4_score,GH 4_adjscore,GH 4_failed,GH 4_missing,GH 5_score,GH 5_adjscore,GH 5_failed,GH 5_missing,GH 6_score,GH 6_adjscore,GH 6_failed,GH 6_missing,GH 7_score,GH 7_adjscore,GH 7_failed,GH 7_missing,GH 8_score,GH 8_adjscore,GH 8_failed,GH 8_missing,GH 9_score,GH 9_adjscore,GH 9_failed,GH 9_missing,GH 10_score,GH 10_adjscore,GH 10_failed,GH 10_missing,GH 11_score,GH 11_adjscore,GH 11_failed,GH 11_missing,GH 12_score,GH 12_adjscore,GH 12_failed,GH 12_missing,GH 13_score,GH 13_adjscore,GH 13_failed,GH 13_missing,GH 14_score,GH 14_adjscore,GH 14_failed,GH 14_missing,GH 15_score,GH 15_adjscore,GH 15_failed,GH 15_missing,GH 17_score,GH 17_adjscore,GH 17_failed,GH 17_missing,GH 19_score,GH 19_adjscore,GH 19_failed,GH 19_missing,GH 21_score,GH 21_adjscore,GH 21_failed,GH 21_missing,GH 22_score,GH 22_adjscore,GH 22_failed,GH 22_missing,GH 24_score,GH 24_adjscore,GH 24_failed,GH 24_missing,GH 25_score,GH 25_adjscore,GH 25_failed,GH 25_missing,GH 27_score,GH 27_adjscore,GH 27_failed,GH 27_missing,GH 28_score,GH 28_adjscore,GH 28_failed,GH 28_missing,GH 29_score,GH 29_adjscore,GH 29_failed,GH 29_missing,GH 30_score,GH 30_adjscore,GH 30_failed,GH 30_missing,GH 31_score,GH 31_adjscore,GH 31_failed,GH 31_missing,GH 32_score,GH 32_adjscore,GH 32_failed,GH 32_missing,GH 33_score,GH 33_adjscore,GH 33_failed,GH 33_missing,GH 34_score,GH 34_adjscore,GH 34_failed,GH 34_missing,GHT_score,GHT_adjscore,GHT_failed,GHT_missing,IF 1_score,IF 1_adjscore,IF 1_failed,IF 1_missing,IF 2_score,IF 2_adjscore,IF 2_failed,IF 2_missing,IF 3_score,IF 3_adjscore,IF 3_failed,IF 3_missing,IF 4_score,IF 4_adjscore,IF 4_failed,IF 4_missing,IF 5_score,IF 5_adjscore,IF 5_failed,IF 5_missing,IF 6_score,IF 6_adjscore,IF 6_failed,IF 6_missing,IFT_score,IFT_adjscore,IFT_failed,IFT_missing,Batch,BWC_Status,Stream_Group,Stream_Unit,target_binary,target_multiclass,target_fighter
0,160CHANC,4.215324,4.215324,False,False,3.937536,3.937536,False,False,4.0,4.0,False,False,3.736023,3.736023,False,False,3.726859,3.726859,False,False,3.7472,3.7472,False,False,3.210893,1.0,True,False,2.673016,1.0,True,False,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,160,Fail,<NA>,<NA>,0,0,0
1,160CHANG,3.926446,3.926446,False,False,3.746286,3.746286,False,False,4.094832,4.094832,False,False,4.036967,4.036967,False,False,4.111887,4.111887,False,False,3.783657,3.783657,False,False,2.528977,2.528977,True,False,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,<NA>,<NA>,<NA>,True,160,Fail,<NA>,<NA>,0,0,0
2,160CHOOC,4.243429,4.243429,False,False,4.37966,4.37966,False,False,4.674136,4.674136,False,False,4.110342,4.110342,False,False,3.962646,3.962646,False,False,3.928343,3.928343,False,False,4.0,4.0,False,False,4.107553,4.107553,False,False,4.304875,4.304875,False,False,4.3375,4.3375,False,False,4.40368

In [18]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 701 entries, 0 to 700
Columns: 156 entries, STUDENT_ID to target_fighter
dtypes: Float64(74), Int64(4), boolean(74), string(4)
memory usage: 615.9 KB


## Target distribution

What the values mean:

- 0: Fail BWC
- 1: Pass BWC, streamed to Fighter
- 2: Pass BWC, streamed to Transport
- 3: Pass BWC, streamed to Heli

The binary target is fairly balanced: 57.06% pass vs 42.94% fail. That means pass/fail modelling is not heavily affected by class imbalance, though a simple majority-class baseline would still get about 57% accuracy.

The multiclass target is more uneven. Fail is the largest class (42.94%), followed by Fighter (28.67%), Rotary (19.40%), and Transport (8.99%). The small Transport class will probably be the hardest to model well, so multiclass evaluation should pay more attention to macro F1 and per-class recall, not just overall accuracy.

At the batch level, there is quite a lot of variation in both pass rates and stream mix. Some batches have much higher pass rates than others, and the proportions of Fighter, Rotary, and Transport also shift across cohorts. This points to possible cohort effects and distribution shift, which could help explain why earlier models did not generalise well to unseen data.

Some of these swings may be due to small batch sizes, so not every difference should be overinterpreted. Still, the overall batch-to-batch variation is large enough to justify batch-aware validation. A random row-wise train/test split would likely give an overly optimistic result, so future-batch or batch-based validation is a better choice.



In [19]:
summarise_col(df, "target_binary")

               count    pct
target_binary              
1                400  57.06
0                301  42.94


### Overall

In [23]:
summarise_col(df, "Stream_Group")
summarise_col(df, "Stream_Unit")
summarise_col(df, "target_binary")
summarise_col(df, "target_multiclass")
summarise_col(df[df["target_multiclass"] != 0],  "target_multiclass")

              count    pct
Stream_Group              
<NA>            301  42.94
Fighter         201  28.67
Heli            136   19.4
Transport        63   8.99
             count    pct
Stream_Unit              
<NA>           301  42.94
RWC            125  17.83
FWC            100  14.27
TWC             63   8.99
SUPT            49   6.99
NFTC            32   4.56
ITAF            20   2.85
IERW            11   1.57
               count    pct
target_binary              
1                400  57.06
0                301  42.94
                   count    pct
target_multiclass              
0                    301  42.94
1                    201  28.67
3                    136   19.4
2                     63   8.99
                   count    pct
target_multiclass              
1                    201  50.25
3                    136   34.0
2                     63  15.75


### By BWC batch

In [24]:
for batch, group in df.groupby("Batch"):
    print(f"BWC Batch {batch}")
    print(f"num students: {len(group)}")
    summarise_col(group, "target_binary")
    summarise_col(group, "Stream_Group")
    print()

BWC Batch 160
num students: 18
               count    pct
target_binary              
1                 10  55.56
0                  8  44.44
              count    pct
Stream_Group              
Fighter           9   50.0
<NA>              8  44.44
Heli              1   5.56

BWC Batch 161
num students: 18
               count    pct
target_binary              
1                 12  66.67
0                  6  33.33
              count    pct
Stream_Group              
Heli              6  33.33
<NA>              6  33.33
Fighter           4  22.22
Transport         2  11.11

BWC Batch 162
num students: 13
               count    pct
target_binary              
1                  8  61.54
0                  5  38.46
              count    pct
Stream_Group              
Fighter           7  53.85
<NA>              5  38.46
Heli              1   7.69

BWC Batch 163
num students: 18
               count    pct
target_binary              
0                 10  55.56
1                  8 

## Feature distribution

In [25]:
def _existing_lesson_feature_cols(
    df: pd.DataFrame,
    lessons: Iterable[str],
    feature_suffixes: Iterable[str] = FEATURE_SUFFIXES,
) -> dict[str, dict[str, str]]:
    """
    Return mapping:
        {
            "GH 1": {
                "score": "GH 1_score",
                "missing": "GH 1_missing",
                ...
            },
            ...
        }
    Only includes columns that actually exist in df.
    """
    lesson_cols: dict[str, dict[str, str]] = {}

    for lesson in lessons:
        feature_map: dict[str, str] = {}
        for suffix in feature_suffixes:
            col = f"{lesson}_{suffix}"
            if col in df.columns:
                feature_map[suffix] = col

        if feature_map:
            lesson_cols[lesson] = feature_map

    return lesson_cols


def _safe_numeric(series: pd.Series) -> pd.Series:
    """Convert to numeric where possible."""
    return pd.to_numeric(series, errors="coerce")


def _safe_bool_rate(series: pd.Series) -> float:
    """
    Return proportion True among all non-missing values.
    If all values are missing, return pd.NA.
    """
    s = series.dropna()
    if s.empty:
        return pd.NA

    # Works for pandas BooleanDtype / bool / object booleans
    return float((s == True).mean())  # noqa: E712


def _safe_value_rate(series: pd.Series, value: object) -> float:
    """
    Return proportion equal to `value` among all non-missing values.
    If all values are missing, return pd.NA.
    """
    s = series.dropna()
    if s.empty:
        return pd.NA
    return float((s == value).mean())


def summarize_key_features_by_col(
    merged: pd.DataFrame,
    lessons_to_keep: list[str],
    batch_col: str = "Batch",
) -> pd.DataFrame:
    """
    Create one long summary table with one row per Batch x Lesson.
    """
    if batch_col not in merged.columns:
        raise KeyError(f"'{batch_col}' not found in dataframe.")

    lesson_feature_map = _existing_lesson_feature_cols(merged, lessons_to_keep)

    if not lesson_feature_map:
        raise ValueError("No lesson feature columns found for the supplied lessons.")

    summary_rows: list[dict[str, object]] = []

    for batch_value, batch_df in merged.groupby(batch_col, dropna=False):
        n_students = len(batch_df)

        for lesson, cols in lesson_feature_map.items():
            row: dict[str, object] = {
                f"{batch_col}": batch_value,
                "lesson": lesson,
                "n_students": n_students,
            }

            # ---- score summary ----
            if "score" in cols:
                s = _safe_numeric(batch_df[cols["score"]])
                row["score_non_null_n"] = int(s.notna().sum())
                row["score_missing_n"] = int(s.isna().sum())
                row["score_mean"] = s.mean()
                row["score_std"] = s.std()
                row["score_median"] = s.median()
                row["score_min"] = s.min()
                row["score_max"] = s.max()

            # ---- adjusted score summary ----
            if "adjscore" in cols:
                s = _safe_numeric(batch_df[cols["adjscore"]])
                row["adjscore_non_null_n"] = int(s.notna().sum())
                row["adjscore_mean"] = s.mean()
                row["adjscore_std"] = s.std()
                row["adjscore_median"] = s.median()
                row["adjscore_min"] = s.min()
                row["adjscore_max"] = s.max()

            # ---- missingness ----
            if "missing" in cols:
                s = batch_df[cols["missing"]]
                row["missing_rate"] = _safe_bool_rate(s)


            # ---- flags ----
            if "failed" in cols:
                row["fail_rate"] = _safe_bool_rate(batch_df[cols["failed"]])

            summary_rows.append(row)

    summary_df = pd.DataFrame(summary_rows).sort_values(
        [f"{batch_col}", "lesson"],
        kind="stable",
    )

    return summary_df.reset_index(drop=True)

### By Stream_Unit

- Highest missingness:
  - NaN: Failures fail at earlier stages, hence high missingness for later scores.
  - With NaN removed, SUPT: High missingness due to SUPT trainees being taken out of BWC earlier.
- Highest fail rate:
  - NaN: The more sorties a trainee fails, the more likely the trainee will fail BWC.
  - With NaN removed, TWC: Supports the notion that the worst-performing trainees get streamed to TWC.
- Highest mean score:
  - SUPT and fighter courses
- Highest mean attempts:
  - IERW dominates: ? People who have to retake tests a lot end up going to heli? but it could be due to Weather

In [26]:
stream_unit_feature_summary = summarize_key_features_by_col(
    merged=df,
    lessons_to_keep=LESSONS_TO_KEEP,
    batch_col="Stream_Unit",
)

print(stream_unit_feature_summary.info())
stream_unit_feature_summary.to_csv("../data/04_EDA/stream_unit/stream_unit_feature_summary.csv", index=False)

<class 'pandas.DataFrame'>
RangeIndex: 296 entries, 0 to 295
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Stream_Unit          259 non-null    str    
 1   lesson               296 non-null    str    
 2   n_students           296 non-null    int64  
 3   score_non_null_n     296 non-null    int64  
 4   score_missing_n      296 non-null    int64  
 5   score_mean           292 non-null    object 
 6   score_std            290 non-null    object 
 7   score_median         292 non-null    object 
 8   score_min            292 non-null    object 
 9   score_max            292 non-null    object 
 10  adjscore_non_null_n  296 non-null    int64  
 11  adjscore_mean        292 non-null    object 
 12  adjscore_std         290 non-null    object 
 13  adjscore_median      292 non-null    object 
 14  adjscore_min         292 non-null    object 
 15  adjscore_max         292 non-null    object 
 16  m

#### Missingness

In [27]:
missing_rate_by_stream_unit = stream_unit_feature_summary.pivot(
    index="lesson",
    columns="Stream_Unit",
    values="missing_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
missing_rate_by_stream_unit = missing_rate_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest missing_rate for each lesson
missing_rate_by_stream_unit["highest_missing"] = missing_rate_by_stream_unit.idxmax(axis=1)

missing_rate_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit/missing_rate_by_stream_unit_lesson.csv"
)

summarise_col(missing_rate_by_stream_unit, "highest_missing")

missing_rate_by_stream_unit

                 count    pct
highest_missing              
NaN                 19  51.35
SUPT                12  32.43
IERW                 3   8.11
ITAF                 3   8.11


Stream_Unit,NaN,FWC,IERW,ITAF,NFTC,RWC,SUPT,TWC,highest_missing
lesson,,,,,,,,,
GH 1,0.262458,0.30,0.181818,0.35,0.18750,0.216,0.387755,0.285714,SUPT
GH 2,0.262458,0.18,0.090909,0.30,0.18750,0.208,0.306122,0.253968,SUPT
GH 3,0.285714,0.24,0.363636,0.10,0.21875,0.184,0.387755,0.158730,SUPT
GH 4,0.318937,0.23,0.363636,0.30,0.18750,0.176,0.285714,0.206349,IERW
GH 5,0.322259,0.21,0.272727,0.30,0.25000,0.256,0.408163,0.301587,SUPT
GH 6,0.358804,0.28,0.272727,0.30,0.12500,0.240,0.224490,0.285714,NaN
GH 7,0.382060,0.28,0.181818,0.20,0.21875,0.272,0.285714,0.317460,NaN
GH 8,0.405316,0.38,0.454545,0.35,0.28125,0.264,0.408163,0.190476,IERW
GH 9,0.358804,0.21,0.181818,0.40,0.12500,0.184,0.265306,0.301587,ITAF


In [28]:
missing_rate_by_stream_unit = stream_unit_feature_summary.dropna(subset="Stream_Unit").pivot(
    index="lesson",
    columns="Stream_Unit",
    values="missing_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
missing_rate_by_stream_unit = missing_rate_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest missing_rate for each lesson
missing_rate_by_stream_unit["highest_missing"] = missing_rate_by_stream_unit.idxmax(axis=1)

missing_rate_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit_no_nans/missing_rate_by_stream_unit_lesson.csv"
)

summarise_col(missing_rate_by_stream_unit, "highest_missing")

missing_rate_by_stream_unit

                 count    pct
highest_missing              
SUPT                22  59.46
IERW                 4  10.81
ITAF                 4  10.81
NFTC                 4  10.81
TWC                  3   8.11


Stream_Unit,FWC,IERW,ITAF,NFTC,RWC,SUPT,TWC,highest_missing
lesson,,,,,,,,
GH 1,0.30,0.181818,0.35,0.18750,0.216,0.387755,0.285714,SUPT
GH 2,0.18,0.090909,0.30,0.18750,0.208,0.306122,0.253968,SUPT
GH 3,0.24,0.363636,0.10,0.21875,0.184,0.387755,0.158730,SUPT
GH 4,0.23,0.363636,0.30,0.18750,0.176,0.285714,0.206349,IERW
GH 5,0.21,0.272727,0.30,0.25000,0.256,0.408163,0.301587,SUPT
GH 6,0.28,0.272727,0.30,0.12500,0.240,0.224490,0.285714,ITAF
GH 7,0.28,0.181818,0.20,0.21875,0.272,0.285714,0.317460,TWC
GH 8,0.38,0.454545,0.35,0.28125,0.264,0.408163,0.190476,IERW
GH 9,0.21,0.181818,0.40,0.12500,0.184,0.265306,0.301587,ITAF


#### Fail rate

In [29]:
fail_rate_by_stream_unit = stream_unit_feature_summary.pivot(
    index="lesson",
    columns="Stream_Unit",
    values="fail_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
fail_rate_by_stream_unit = fail_rate_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest fail_rate for each lesson
fail_rate_by_stream_unit["highest_fail_rate"] = fail_rate_by_stream_unit.idxmax(axis=1)

fail_rate_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit/fail_rate_by_stream_unit_lesson.csv"
)

summarise_col(fail_rate_by_stream_unit, "highest_fail_rate")

fail_rate_by_stream_unit

                   count    pct
highest_fail_rate              
NaN                   16  43.24
TWC                    9  24.32
IERW                   8  21.62
RWC                    1   2.70
NFTC                   1   2.70
ITAF                   1   2.70
SUPT                   1   2.70


Stream_Unit,NaN,FWC,IERW,ITAF,NFTC,RWC,SUPT,TWC,highest_fail_rate
lesson,,,,,,,,,
GH 1,0.009009,0.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN
GH 2,0.018018,0.012195,0.0,0.0,0.0,0.020202,0.0,0.0,RWC
GH 3,0.018605,0.0,0.0,0.0,0.04,0.0,0.0,0.0,NFTC
GH 4,0.043902,0.012987,0.0,0.0,0.0,0.029126,0.0,0.0,NaN
GH 5,0.063725,0.063291,0.0,0.0,0.0,0.032258,0.0,0.090909,TWC
GH 6,0.11399,0.027778,0.125,0.0,0.0,0.021053,0.026316,0.066667,IERW
GH 7,0.080645,0.055556,0.111111,0.0,0.08,0.076923,0.028571,0.046512,IERW
GH 8,0.184358,0.032258,0.0,0.0,0.086957,0.152174,0.137931,0.098039,NaN
GH 9,0.238342,0.075949,0.111111,0.166667,0.071429,0.117647,0.0,0.181818,NaN


In [30]:
fail_rate_by_stream_unit = stream_unit_feature_summary.dropna(subset="Stream_Unit").pivot(
    index="lesson",
    columns="Stream_Unit",
    values="fail_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
fail_rate_by_stream_unit = fail_rate_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest fail_rate for each lesson
fail_rate_by_stream_unit["highest_fail_rate"] = fail_rate_by_stream_unit.idxmax(axis=1)

fail_rate_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit_no_nans/fail_rate_by_stream_unit_lesson.csv"
)

summarise_col(fail_rate_by_stream_unit, "highest_fail_rate")

fail_rate_by_stream_unit

                   count    pct
highest_fail_rate              
TWC                   14  37.84
IERW                  11  29.73
FWC                    3   8.11
RWC                    3   8.11
NFTC                   3   8.11
ITAF                   2   5.41
SUPT                   1   2.70


Stream_Unit,FWC,IERW,ITAF,NFTC,RWC,SUPT,TWC,highest_fail_rate
lesson,,,,,,,,
GH 1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,FWC
GH 2,0.012195,0.0,0.0,0.0,0.020202,0.0,0.0,RWC
GH 3,0.0,0.0,0.0,0.04,0.0,0.0,0.0,NFTC
GH 4,0.012987,0.0,0.0,0.0,0.029126,0.0,0.0,RWC
GH 5,0.063291,0.0,0.0,0.0,0.032258,0.0,0.090909,TWC
GH 6,0.027778,0.125,0.0,0.0,0.021053,0.026316,0.066667,IERW
GH 7,0.055556,0.111111,0.0,0.08,0.076923,0.028571,0.046512,IERW
GH 8,0.032258,0.0,0.0,0.086957,0.152174,0.137931,0.098039,RWC
GH 9,0.075949,0.111111,0.166667,0.071429,0.117647,0.0,0.181818,TWC


#### score_mean

In [31]:
score_mean_by_stream_unit = stream_unit_feature_summary.pivot(
    index="lesson",
    columns="Stream_Unit",
    values="score_mean",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
score_mean_by_stream_unit = score_mean_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest score_mean for each lesson
score_mean_by_stream_unit["highest_score"] = score_mean_by_stream_unit.idxmax(axis=1)

score_mean_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit/score_mean_by_stream_unit_lesson.csv"
)

summarise_col(score_mean_by_stream_unit, "highest_score")

score_mean_by_stream_unit

               count    pct
highest_score              
SUPT              22  59.46
ITAF              10  27.03
FWC                2   5.41
NFTC               2   5.41
IERW               1   2.70


Stream_Unit,NaN,FWC,IERW,ITAF,NFTC,RWC,SUPT,TWC,highest_score
lesson,,,,,,,,,
GH 1,4.692296,4.949411,5.077998,5.268062,4.736339,4.802758,4.922199,4.849831,ITAF
GH 2,4.629744,5.056437,5.121626,5.064885,4.791849,4.760327,5.018525,4.804052,IERW
GH 3,4.376686,4.731431,4.317469,4.758475,4.548008,4.463989,4.769112,4.484887,SUPT
GH 4,4.370463,4.676477,4.441278,4.786229,4.530769,4.476667,4.688946,4.504464,ITAF
GH 5,4.214816,4.502656,4.592361,4.779708,4.578382,4.339215,4.655045,4.360977,ITAF
GH 6,4.162812,4.493043,4.32866,4.590176,4.238308,4.319181,4.435927,4.324963,ITAF
GH 7,4.196908,4.474836,4.386774,4.812234,4.285744,4.2743,4.57638,4.342752,ITAF
GH 8,4.108857,4.38431,4.038756,4.707248,4.246888,4.269862,4.405617,4.349025,ITAF
GH 9,4.083597,4.326781,4.220264,4.539312,4.309932,4.254437,4.564584,4.171841,SUPT


In [32]:
score_mean_by_stream_unit = stream_unit_feature_summary.dropna(subset="Stream_Unit").pivot(
    index="lesson",
    columns="Stream_Unit",
    values="score_mean",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
score_mean_by_stream_unit = score_mean_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest score_mean for each lesson
score_mean_by_stream_unit["highest_score"] = score_mean_by_stream_unit.idxmax(axis=1)

score_mean_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit_no_nans/score_mean_by_stream_unit_lesson.csv"
)

summarise_col(score_mean_by_stream_unit, "highest_score")

score_mean_by_stream_unit

               count    pct
highest_score              
SUPT              22  59.46
ITAF              10  27.03
FWC                2   5.41
NFTC               2   5.41
IERW               1   2.70


Stream_Unit,FWC,IERW,ITAF,NFTC,RWC,SUPT,TWC,highest_score
lesson,,,,,,,,
GH 1,4.949411,5.077998,5.268062,4.736339,4.802758,4.922199,4.849831,ITAF
GH 2,5.056437,5.121626,5.064885,4.791849,4.760327,5.018525,4.804052,IERW
GH 3,4.731431,4.317469,4.758475,4.548008,4.463989,4.769112,4.484887,SUPT
GH 4,4.676477,4.441278,4.786229,4.530769,4.476667,4.688946,4.504464,ITAF
GH 5,4.502656,4.592361,4.779708,4.578382,4.339215,4.655045,4.360977,ITAF
GH 6,4.493043,4.32866,4.590176,4.238308,4.319181,4.435927,4.324963,ITAF
GH 7,4.474836,4.386774,4.812234,4.285744,4.2743,4.57638,4.342752,ITAF
GH 8,4.38431,4.038756,4.707248,4.246888,4.269862,4.405617,4.349025,ITAF
GH 9,4.326781,4.220264,4.539312,4.309932,4.254437,4.564584,4.171841,SUPT


#### mean_attempts

In [ ]:
mean_attempts_by_stream_unit = stream_unit_feature_summary.pivot(
    index="lesson",
    columns="Stream_Unit",
    values="mean_attempts",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
mean_attempts_by_stream_unit = mean_attempts_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest mean_attempts for each lesson
mean_attempts_by_stream_unit["highest_mean_attempts"] = mean_attempts_by_stream_unit.idxmax(axis=1)

mean_attempts_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit/mean_attempts_by_stream_unit_lesson.csv"
)

summarise_col(mean_attempts_by_stream_unit, "highest_mean_attempts")

mean_attempts_by_stream_unit

In [ ]:
mean_attempts_by_stream_unit = stream_unit_feature_summary.dropna(subset="Stream_Unit").pivot(
    index="lesson",
    columns="Stream_Unit",
    values="mean_attempts",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
mean_attempts_by_stream_unit = mean_attempts_by_stream_unit.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest mean_attempts for each lesson
mean_attempts_by_stream_unit["highest_mean_attempts"] = mean_attempts_by_stream_unit.idxmax(axis=1)

mean_attempts_by_stream_unit.to_csv(
    "../data/04_EDA/stream_unit_no_nans/mean_attempts_by_stream_unit_lesson.csv"
)

summarise_col(mean_attempts_by_stream_unit, "highest_mean_attempts")

mean_attempts_by_stream_unit

#### Compact

In [34]:
compact_stream_unit_feature_summary = stream_unit_feature_summary[
    [
        "Stream_Unit",
        "lesson",
        "n_students",
        "score_mean",
        "score_median",
        "missing_rate",
        "fail_rate",
    ]
].copy()
compact_stream_unit_feature_summary.to_csv("../data/04_EDA/stream_unit/compact_stream_unit_feature_summary.csv", index=False)
compact_stream_unit_feature_summary.info()

<class 'pandas.DataFrame'>
RangeIndex: 296 entries, 0 to 295
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Stream_Unit   259 non-null    str    
 1   lesson        296 non-null    str    
 2   n_students    296 non-null    int64  
 3   score_mean    292 non-null    object 
 4   score_median  292 non-null    object 
 5   missing_rate  296 non-null    float64
 6   fail_rate     292 non-null    object 
dtypes: float64(1), int64(1), object(3), str(2)
memory usage: 18.6+ KB


### By Stream_Group

- Highest missingness:
  - NaN: Failures fail at earlier stages, hence high missingness for later scores.
  - With NaN removed, Fighter: High missingness due to SUPT.
- Highest fail rate:
  - NaN: The more sorties a trainee fails, the more likely the trainee will fail BWC.
  - With NaN removed, Transport: Supports the notion that the worst-performing trainees get streamed to Transport.
  - Supported by Transport having the highest DNCO and NG rates.
- Highest mean score:
  - Fighter: supports the notion that the best-performing trainees get streamed to Fighter
- Lowest mean score:
  - NaN: Trainees who achieve low scores fail BWC
  - With NaN removed, even split between heli and transport
- Highest mean attempts:
  - Heli: ? People who have to retake tests a lot end up going to heli? but it could be due to Weather

In [35]:
stream_group_feature_summary = summarize_key_features_by_col(
    merged=df,
    lessons_to_keep=LESSONS_TO_KEEP,
    batch_col="Stream_Group",
)

print(stream_group_feature_summary.info())
stream_group_feature_summary.to_csv("../data/04_EDA/stream_group/stream_group_feature_summary.csv", index=False)

<class 'pandas.DataFrame'>
RangeIndex: 148 entries, 0 to 147
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Stream_Group         111 non-null    str    
 1   lesson               148 non-null    str    
 2   n_students           148 non-null    int64  
 3   score_non_null_n     148 non-null    int64  
 4   score_missing_n      148 non-null    int64  
 5   score_mean           148 non-null    float64
 6   score_std            148 non-null    float64
 7   score_median         148 non-null    float64
 8   score_min            148 non-null    float64
 9   score_max            148 non-null    float64
 10  adjscore_non_null_n  148 non-null    int64  
 11  adjscore_mean        148 non-null    float64
 12  adjscore_std         148 non-null    float64
 13  adjscore_median      148 non-null    float64
 14  adjscore_min         148 non-null    float64
 15  adjscore_max         148 non-null    float64
 16  m

#### Missingness

In [36]:
missing_rate_by_stream_group = stream_group_feature_summary.pivot(
    index="lesson",
    columns="Stream_Group",
    values="missing_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
missing_rate_by_stream_group = missing_rate_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest missing_rate for each lesson
missing_rate_by_stream_group["highest_missing"] = missing_rate_by_stream_group.idxmax(axis=1)

missing_rate_by_stream_group.to_csv(
    "../data/04_EDA/stream_group/missing_rate_by_stream_group_lesson.csv"
)

summarise_col(missing_rate_by_stream_group, "highest_missing")

missing_rate_by_stream_group

                 count   pct
highest_missing             
NaN                 36  97.3
Fighter              1   2.7


Stream_Group,NaN,Fighter,Heli,Transport,highest_missing
lesson,,,,,
GH 1,0.262458,0.308458,0.213235,0.285714,Fighter
GH 2,0.262458,0.223881,0.198529,0.253968,NaN
GH 3,0.285714,0.258706,0.198529,0.158730,NaN
GH 4,0.318937,0.243781,0.191176,0.206349,NaN
GH 5,0.322259,0.273632,0.257353,0.301587,NaN
GH 6,0.358804,0.243781,0.242647,0.285714,NaN
GH 7,0.382060,0.263682,0.264706,0.317460,NaN
GH 8,0.405316,0.368159,0.279412,0.190476,NaN
GH 9,0.358804,0.228856,0.183824,0.301587,NaN


In [37]:
missing_rate_by_stream_group = stream_group_feature_summary.dropna(subset="Stream_Group").pivot(
    index="lesson",
    columns="Stream_Group",
    values="missing_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
missing_rate_by_stream_group = missing_rate_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest missing_rate for each lesson
missing_rate_by_stream_group["highest_missing"] = missing_rate_by_stream_group.idxmax(axis=1)

missing_rate_by_stream_group.to_csv(
    "../data/04_EDA/stream_group_no_nans/missing_rate_by_stream_group_lesson.csv"
)

summarise_col(missing_rate_by_stream_group, "highest_missing")

missing_rate_by_stream_group

                 count    pct
highest_missing              
Transport           19  51.35
Fighter             16  43.24
Heli                 2   5.41


Stream_Group,Fighter,Heli,Transport,highest_missing
lesson,,,,
GH 1,0.308458,0.213235,0.285714,Fighter
GH 2,0.223881,0.198529,0.253968,Transport
GH 3,0.258706,0.198529,0.158730,Fighter
GH 4,0.243781,0.191176,0.206349,Fighter
GH 5,0.273632,0.257353,0.301587,Transport
GH 6,0.243781,0.242647,0.285714,Transport
GH 7,0.263682,0.264706,0.317460,Transport
GH 8,0.368159,0.279412,0.190476,Fighter
GH 9,0.228856,0.183824,0.301587,Transport


#### Fail rate

In [38]:
fail_rate_by_stream_group = stream_group_feature_summary.pivot(
    index="lesson",
    columns="Stream_Group",
    values="fail_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
fail_rate_by_stream_group = fail_rate_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest fail_rate for each lesson
fail_rate_by_stream_group["highest_fail_rate"] = fail_rate_by_stream_group.idxmax(axis=1)

fail_rate_by_stream_group.to_csv(
    "../data/04_EDA/stream_group/fail_rate_by_stream_group_lesson.csv"
)

summarise_col(fail_rate_by_stream_group, "highest_fail_rate")

fail_rate_by_stream_group

                   count    pct
highest_fail_rate              
NaN                   23  62.16
Transport             13  35.14
Heli                   1   2.70


Stream_Group,NaN,Fighter,Heli,Transport,highest_fail_rate
lesson,,,,,
GH 1,0.009009,0.000000,0.000000,0.000000,NaN
GH 2,0.018018,0.006410,0.018349,0.000000,Heli
GH 3,0.018605,0.006711,0.000000,0.000000,NaN
GH 4,0.043902,0.006579,0.027273,0.000000,NaN
GH 5,0.063725,0.034247,0.029703,0.090909,Transport
GH 6,0.113990,0.019737,0.029126,0.066667,NaN
GH 7,0.080645,0.047297,0.080000,0.046512,NaN
GH 8,0.184358,0.062992,0.142857,0.098039,NaN
GH 9,0.238342,0.064516,0.117117,0.181818,NaN


In [39]:
fail_rate_by_stream_group = stream_group_feature_summary.dropna(subset="Stream_Group").pivot(
    index="lesson",
    columns="Stream_Group",
    values="fail_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
fail_rate_by_stream_group = fail_rate_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest fail_rate for each lesson
fail_rate_by_stream_group["highest_fail_rate"] = fail_rate_by_stream_group.idxmax(axis=1)

fail_rate_by_stream_group.to_csv(
    "../data/04_EDA/stream_group_no_nans/fail_rate_by_stream_group_lesson.csv"
)

summarise_col(fail_rate_by_stream_group, "highest_fail_rate")

fail_rate_by_stream_group

                   count    pct
highest_fail_rate              
Transport             24  64.86
Heli                   7  18.92
Fighter                6  16.22


Stream_Group,Fighter,Heli,Transport,highest_fail_rate
lesson,,,,
GH 1,0.000000,0.000000,0.000000,Fighter
GH 2,0.006410,0.018349,0.000000,Heli
GH 3,0.006711,0.000000,0.000000,Fighter
GH 4,0.006579,0.027273,0.000000,Heli
GH 5,0.034247,0.029703,0.090909,Transport
GH 6,0.019737,0.029126,0.066667,Transport
GH 7,0.047297,0.080000,0.046512,Heli
GH 8,0.062992,0.142857,0.098039,Heli
GH 9,0.064516,0.117117,0.181818,Transport


#### score_mean

In [40]:
score_mean_by_stream_group = stream_group_feature_summary.pivot(
    index="lesson",
    columns="Stream_Group",
    values="score_mean",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
score_mean_by_stream_group = score_mean_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest score_mean for each lesson
score_mean_by_stream_group["highest_score"] = score_mean_by_stream_group.idxmax(axis=1)

# Get Stream Unit with the lowest score_mean for each lesson
numeric_cols = score_mean_by_stream_group.select_dtypes(include="number").columns

score_mean_by_stream_group["lowest_score"] = (
    score_mean_by_stream_group[numeric_cols].idxmin(axis=1)
)

score_mean_by_stream_group.to_csv(
    "../data/04_EDA/stream_group/score_mean_by_stream_group_lesson.csv"
)

summarise_col(score_mean_by_stream_group, "highest_score")
summarise_col(score_mean_by_stream_group, "lowest_score")

score_mean_by_stream_group

               count   pct
highest_score             
Fighter           36  97.3
NaN                1   2.7
              count    pct
lowest_score              
NaN              23  62.16
Transport        10  27.03
Heli              4  10.81


Stream_Group,NaN,Fighter,Heli,Transport,highest_score,lowest_score
lesson,,,,,,
GH 1,4.692296,4.933484,4.825909,4.849831,Fighter,NaN
GH 2,4.629744,5.004834,4.793473,4.804052,Fighter,NaN
GH 3,4.376686,4.711509,4.454579,4.484887,Fighter,NaN
GH 4,4.370463,4.664533,4.474415,4.504464,Fighter,NaN
GH 5,4.214816,4.571940,4.359266,4.360977,Fighter,NaN
GH 6,4.162812,4.440786,4.319917,4.324963,Fighter,NaN
GH 7,4.196908,4.503384,4.284423,4.342752,Fighter,NaN
GH 8,4.108857,4.397345,4.255713,4.349025,Fighter,NaN
GH 9,4.083597,4.395423,4.251666,4.171841,Fighter,NaN


In [41]:
score_mean_by_stream_group = stream_group_feature_summary.dropna(subset="Stream_Group").pivot(
    index="lesson",
    columns="Stream_Group",
    values="score_mean",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
score_mean_by_stream_group = score_mean_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest score_mean for each lesson
score_mean_by_stream_group["highest_score"] = score_mean_by_stream_group.idxmax(axis=1)

# Get Stream Unit with the lowest score_mean for each lesson
numeric_cols = score_mean_by_stream_group.select_dtypes(include="number").columns

score_mean_by_stream_group["lowest_score"] = (
    score_mean_by_stream_group[numeric_cols].idxmin(axis=1)
)

score_mean_by_stream_group.to_csv(
    "../data/04_EDA/stream_group_no_nans/score_mean_by_stream_group_lesson.csv"
)

summarise_col(score_mean_by_stream_group, "highest_score")
summarise_col(score_mean_by_stream_group, "lowest_score")

score_mean_by_stream_group

               count    pct
highest_score              
Fighter           37  100.0
              count    pct
lowest_score              
Transport        23  62.16
Heli             14  37.84


Stream_Group,Fighter,Heli,Transport,highest_score,lowest_score
lesson,,,,,
GH 1,4.933484,4.825909,4.849831,Fighter,Heli
GH 2,5.004834,4.793473,4.804052,Fighter,Heli
GH 3,4.711509,4.454579,4.484887,Fighter,Heli
GH 4,4.664533,4.474415,4.504464,Fighter,Heli
GH 5,4.571940,4.359266,4.360977,Fighter,Heli
GH 6,4.440786,4.319917,4.324963,Fighter,Heli
GH 7,4.503384,4.284423,4.342752,Fighter,Heli
GH 8,4.397345,4.255713,4.349025,Fighter,Heli
GH 9,4.395423,4.251666,4.171841,Fighter,Transport


#### adjscore_mean

In [42]:
adjscore_mean_by_stream_group = stream_group_feature_summary.pivot(
    index="lesson",
    columns="Stream_Group",
    values="adjscore_mean",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
adjscore_mean_by_stream_group = adjscore_mean_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest adjscore_mean for each lesson
adjscore_mean_by_stream_group["highest_score"] = adjscore_mean_by_stream_group.idxmax(axis=1)

# Get Stream Unit with the lowest adjscore_mean for each lesson
numeric_cols = adjscore_mean_by_stream_group.select_dtypes(include="number").columns

adjscore_mean_by_stream_group["lowest_score"] = (
    adjscore_mean_by_stream_group[numeric_cols].idxmin(axis=1)
)

adjscore_mean_by_stream_group.to_csv(
    "../data/04_EDA/stream_group/adjscore_mean_by_stream_group_lesson.csv"
)

summarise_col(adjscore_mean_by_stream_group, "highest_score")
summarise_col(adjscore_mean_by_stream_group, "lowest_score")

adjscore_mean_by_stream_group

               count    pct
highest_score              
Fighter           35  94.59
Transport          1   2.70
NaN                1   2.70
              count    pct
lowest_score              
NaN              24  64.86
Transport         8  21.62
Heli              5  13.51


Stream_Group,NaN,Fighter,Heli,Transport,highest_score,lowest_score
lesson,,,,,,
GH 1,4.678203,4.933484,4.825909,4.849831,Fighter,NaN
GH 2,4.604177,4.985875,4.789102,4.804052,Fighter,NaN
GH 3,4.351243,4.711509,4.443416,4.472758,Fighter,NaN
GH 4,4.316573,4.656039,4.474415,4.504464,Fighter,NaN
GH 5,4.089426,4.516425,4.333274,4.360977,Fighter,NaN
GH 6,4.021538,4.356059,4.309830,4.324963,Fighter,NaN
GH 7,4.046781,4.471931,4.143675,4.194831,Fighter,NaN
GH 8,3.978307,4.363080,4.227524,4.290974,Fighter,NaN
GH 9,3.992754,4.343182,4.251666,4.099548,Fighter,NaN


In [43]:
adjscore_mean_by_stream_group = stream_group_feature_summary.dropna(subset="Stream_Group").pivot(
    index="lesson",
    columns="Stream_Group",
    values="adjscore_mean",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
adjscore_mean_by_stream_group = adjscore_mean_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest adjscore_mean for each lesson
adjscore_mean_by_stream_group["highest_score"] = adjscore_mean_by_stream_group.idxmax(axis=1)

# Get Stream Unit with the lowest adjscore_mean for each lesson
numeric_cols = adjscore_mean_by_stream_group.select_dtypes(include="number").columns

adjscore_mean_by_stream_group["lowest_score"] = (
    adjscore_mean_by_stream_group[numeric_cols].idxmin(axis=1)
)

adjscore_mean_by_stream_group.to_csv(
    "../data/04_EDA/stream_group_no_nans/adjscore_mean_by_stream_group_lesson.csv"
)

summarise_col(adjscore_mean_by_stream_group, "highest_score")
summarise_col(adjscore_mean_by_stream_group, "lowest_score")

adjscore_mean_by_stream_group

               count   pct
highest_score             
Fighter           36  97.3
Transport          1   2.7
              count    pct
lowest_score              
Transport        19  51.35
Heli             17  45.95
Fighter           1   2.70


Stream_Group,Fighter,Heli,Transport,highest_score,lowest_score
lesson,,,,,
GH 1,4.933484,4.825909,4.849831,Fighter,Heli
GH 2,4.985875,4.789102,4.804052,Fighter,Heli
GH 3,4.711509,4.443416,4.472758,Fighter,Heli
GH 4,4.656039,4.474415,4.504464,Fighter,Heli
GH 5,4.516425,4.333274,4.360977,Fighter,Heli
GH 6,4.356059,4.309830,4.324963,Fighter,Heli
GH 7,4.471931,4.143675,4.194831,Fighter,Heli
GH 8,4.363080,4.227524,4.290974,Fighter,Heli
GH 9,4.343182,4.251666,4.099548,Fighter,Transport


#### mean_attempts

In [ ]:
mean_attempts_by_stream_group = stream_group_feature_summary.pivot(
    index="lesson",
    columns="Stream_Group",
    values="mean_attempts",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
mean_attempts_by_stream_group = mean_attempts_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest mean_attempts for each lesson
mean_attempts_by_stream_group["highest_mean_attempts"] = mean_attempts_by_stream_group.idxmax(axis=1)

mean_attempts_by_stream_group.to_csv(
    "../data/04_EDA/stream_group/mean_attempts_by_stream_group_lesson.csv"
)

summarise_col(mean_attempts_by_stream_group, "highest_mean_attempts")

mean_attempts_by_stream_group

In [ ]:
mean_attempts_by_stream_group = stream_group_feature_summary.dropna(subset="Stream_Group").pivot(
    index="lesson",
    columns="Stream_Group",
    values="mean_attempts",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
mean_attempts_by_stream_group = mean_attempts_by_stream_group.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest mean_attempts for each lesson
mean_attempts_by_stream_group["highest_mean_attempts"] = mean_attempts_by_stream_group.idxmax(axis=1)

mean_attempts_by_stream_group.to_csv(
    "../data/04_EDA/stream_group_no_nans/mean_attempts_by_stream_group_lesson.csv"
)

summarise_col(mean_attempts_by_stream_group, "highest_mean_attempts")

mean_attempts_by_stream_group

#### Compact

In [44]:
compact_stream_group_feature_summary = stream_group_feature_summary[
    [
        "Stream_Group",
        "lesson",
        "n_students",
        "score_mean",
        "score_median",
        "missing_rate",
        "fail_rate",
    ]
].copy()
compact_stream_group_feature_summary.to_csv("../data/04_EDA/stream_group/compact_stream_group_feature_summary.csv", index=False)
compact_stream_group_feature_summary.info()

<class 'pandas.DataFrame'>
RangeIndex: 148 entries, 0 to 147
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Stream_Group  111 non-null    str    
 1   lesson        148 non-null    str    
 2   n_students    148 non-null    int64  
 3   score_mean    148 non-null    float64
 4   score_median  148 non-null    float64
 5   missing_rate  148 non-null    float64
 6   fail_rate     148 non-null    float64
dtypes: float64(4), int64(1), str(2)
memory usage: 9.6 KB


### By Batch

- Missingness:
  - Some lessons were only introduced later: e.g. GH 30 and GH 32 are essentially 100% missing from batches 160–185, then become meaningfully populated from batch 186 onward. Or the lesson name changed and I failed to capture the original one?
  - Some lessons seem to have been removed in later batches: e.g. GH 31 from batch 186 onwards is essentially 100% missing
  - IF6 missing from 168 to 185
  - later lessons have higher missingness

In [46]:
batch_feature_summary = summarize_key_features_by_col(
    merged=df,
    lessons_to_keep=LESSONS_TO_KEEP,
    batch_col="Batch",
)

print(batch_feature_summary.info())
batch_feature_summary.to_csv("../data/04_EDA/batch/batch_feature_summary.csv", index=False)

<class 'pandas.DataFrame'>
RangeIndex: 1739 entries, 0 to 1738
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Batch                1739 non-null   int64  
 1   lesson               1739 non-null   str    
 2   n_students           1739 non-null   int64  
 3   score_non_null_n     1739 non-null   int64  
 4   score_missing_n      1739 non-null   int64  
 5   score_mean           1566 non-null   object 
 6   score_std            1551 non-null   object 
 7   score_median         1566 non-null   object 
 8   score_min            1566 non-null   object 
 9   score_max            1566 non-null   object 
 10  adjscore_non_null_n  1739 non-null   int64  
 11  adjscore_mean        1566 non-null   object 
 12  adjscore_std         1551 non-null   object 
 13  adjscore_median      1566 non-null   object 
 14  adjscore_min         1566 non-null   object 
 15  adjscore_max         1566 non-null   object 
 16 

#### Missingness

In [47]:
missing_rate_by_batch = batch_feature_summary.pivot(
    index="lesson",
    columns="Batch",
    values="missing_rate",
)

# sort the lesson index naturally: GH 1, GH 2, ..., GH 10
missing_rate_by_batch = missing_rate_by_batch.sort_index(
    key=lambda idx: idx.map(natural_sort_key)
)

# Get Stream Unit with the highest missing_rate for each lesson
missing_rate_by_batch["highest_missing"] = missing_rate_by_batch.idxmax(axis=1)

missing_rate_by_batch.to_csv(
    "../data/04_EDA/batch/missing_rate_by_batch_lesson.csv"
)

summarise_col(missing_rate_by_batch, "highest_missing")

missing_rate_by_batch

                 count    pct
highest_missing              
206                 28  75.68
160                  2   5.41
190                  1   2.70
170                  1   2.70
186                  1   2.70
162                  1   2.70
171                  1   2.70
187                  1   2.70
168                  1   2.70


Batch,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,highest_missing
lesson,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
GH 1,0.000000,0.000000,0.076923,0.111111,0.235294,0.3750,0.111111,0.117647,0.142857,0.105263,0.133333,0.000000,0.10,0.833333,0.363636,0.083333,0.05,0.214286,0.076923,0.000000,0.000000,0.142857,0.333333,0.294118,0.250000,0.500000,0.3750,0.473684,0.461538,0.384615,0.444444,0.583333,0.235294,0.529412,0.4375,0.176471,0.176471,0.4375,0.214286,0.333333,0.5625,0.545455,0.2500,0.571429,0.666667,0.444444,1.0,206
GH 2,0.055556,0.000000,0.000000,0.222222,0.235294,0.3125,0.222222,0.000000,0.000000,0.263158,0.066667,0.142857,0.15,0.500000,0.090909,0.000000,0.00,0.071429,0.153846,0.222222,0.200000,0.142857,0.111111,0.235294,0.250000,0.277778,0.2500,0.368421,0.230769,0.461538,0.444444,0.250000,0.470588,0.235294,0.1250,0.000000,0.352941,0.5625,0.357143,0.533333,0.5625,0.818182,0.3750,0.000000,0.500000,0.222222,1.0,206
GH 3,0.055556,0.000000,0.000000,0.277778,0.176471,0.2500,0.333333,0.235294,0.214286,0.263158,0.066667,0.095238,0.25,0.583333,0.272727,0.083333,0.00,0.071429,0.076923,0.111111,0.266667,0.000000,0.222222,0.117647,0.250000,0.333333,0.1875,0.157895,0.461538,0.384615,0.611111,0.250000,0.470588,0.294118,0.3125,0.352941,0.176471,0.3750,0.071429,0.333333,0.3750,0.727273,0.3125,0.571429,0.500000,0.333333,1.0,206
GH 4,0.000000,0.000000,0.000000,0.222222,0.176471,0.1875,0.111111,0.235294,0.142857,0.263158,0.066667,0.238095,0.25,0.250000,0.363636,0.000000,0.05,0.071429,0.000000,0.000000,0.200000,0.071429,0.166667,0.176471,0.333333,0.444444,0.6250,0.263158,0.230769,0.384615,0.500000,0.416667,0.352941,0.529412,0.3125,0.352941,0.235294,0.1875,0.428571,0.600000,0.3750,0.818182,0.1875,0.714286,0.416667,0.555556,1.0,206
GH 5,0.055556,0.000000,0.076923,0.333333,0.117647,0.5625,0.222222,0.117647,0.142857,0.210526,0.133333,0.190476,0.30,0.500000,0.090909,0.083333,0.05,0.142857,0.000000,0.000000,0.133333,0.142857,0.055556,0.470588,0.500000,0.444444,0.3750,0.631579,0.307692,0.846154,0.500000,0.083333,0.588235,0.411765,0.1875,0.294118,0.235294,0.3750,0.500000,0.200000,0.5000,0.909091,0.1875,0.857143,0.416667,0.444444,1.0,206
GH 6,0.000000,0.111111,0.000000,0.277778,0.000000,0.3125,0.166667,0.117647,0.214286,0.315789,0.133333,0.571429,0.30,0.500000,0.090909,0.000000,0.20,0.000000,0.076923,0.000000,0.200000,0.071429,0.333333,0.411765,0.666667,0.555556,0.3125,0.526316,0.153846,0.615385,0.722222,0.166667,0.352941,0.235294,0.4375,0.352941,0.411765,0.1875,0.500000,0.600000,0.1250,0.545455,0.2500,0.428571,0.333333,0.666667,1.0,206
GH 7,0.055556,0.111111,0.000000,0.500000,0.176471,0.3750,0.388889,0.294118,0.142857,0.421053,0.266667,0.428571,0.30,0.416667,0.272727,0.083333,0.20,0.071429,0.000000,0.111111,0.133333,0.285714,0.166667,0.294118,0.333333,0.500000,0.4375,0.368421,0.230769,0.615385,0.888889,0.166667,0.294118,0.235294,0.5625,0.352941,0.235294,0.4375,0.285714,0.533333,0.4375,0.909091,0.0625,0.714286,0.166667,0.444444,1.0,206
GH 8,0.166667,0.166667,0.076923,0.333333,0.176471,0.5625,0.277778,0.176471,0.142857,0.210526,0.333333,0.333333,0.20,0.333333,0.181818,0.166667,0.40,0.142857,0.153846,0.333333,0.266667,0.357143,0.277778,0.117647,0.416667,0.666667,0.5000,0.157895,0.461538,0.615385,1.000000,0.083333,0.411765,0.235294,0.7500,0.235294,0.470588,0.4375,0.571429,0.733333,0.1875,0.636364,0.1875,0.571429,0.416667,0.777778,1.0,190
GH 9,0.222222,0.166667,0.000000,0.055556,0.176471,0.3125,0.055556,0.058824,0.214286,0.052632,0.200000,0.190476,0.20,0.500000,0.454545,0.333333,0.20,0.214286,0.153846,0.444444,0.066667,0.000000,0.222222,0.529412,0.500000,0.611111,0.0000,0.315789,0.153846,0.538462,0.777778,0.416667,0.470588,0.294118,0.5000,0.352941,0.352941,0.3125,0.571429,0.466667,0.3125,0.272727,0.1875,0.142857,0.083333,0.555556,1.0,206
